aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+    auto *ptr = data + index * stride;
+    return Vec256<scalar_t>::loadu(ptr);
+  }
+};
+
+template <typename T>
+T load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+  return LoadImpl<T>::load(data, stride, index);
+}
+
+template <typename scalar_t>
+void store_result(char * C10_RESTRICT data, int64_t stride, int64_t index, scalar_t value) {
+  auto * ptr = reinterpret_cast<scalar_t*>(data + index * stride);
+  *ptr += value;
+}
+
+template <typename scalar_t, size_t numel>
+void store_result(char * C10_RESTRICT data, int64_t stride, int64_t index,
+    const std::array<scalar_t, numel> &values) {
+  auto *base_ptr = data + stride * index;
+  for (int64_t k = 0; k < numel; ++k) {
+    store_result(base_ptr, stride, k, values[k]);
+  }
+}
+
+int64_t ceil_log2(int64_t x) {
+  if (x <= 2) {
+    return 1;
+  }
+
+  auto ux = static_cast<uint64_t>(x);
+  // Last set bit is floor(log2(x)), floor + 1 is ceil
+  // except when x is an exact powers of 2, so subtract 1 first
+  return static_cast<int64_t>(llvm::findLastSet(ux - 1)) + 1;
+}
+
+// Simultaneously sum over n rows at once
+template <typename scalar_t, int64_t nrows>
+std::array<scalar_t, nrows> multi_row_sum(
+    const char * C10_RESTRICT in_data,
+    const int64_t row_stride,
+    const int64_t col_stride,
+    const int64_t size) {
+  constexpr int64_t num_levels = 4;
+
+  const int64_t level_power =
+      std::max(int64_t(4), ceil_log2(size) / num_levels);
+  const int64_t level_step = (1 << level_power);
+  const int64_t level_mask = level_step - 1;
+
+  scalar_t acc[num_levels][nrows];
+  std::fill_n(&acc[0][0], num_levels * nrows, scalar_t(0));
+
+  int64_t i = 0;
+  for (; i + level_step <= size;) {
+    for (int64_t j = 0; j < level_step; ++j, ++i) {
+      const char * sum_base = in_data + i * row_stride;
+      #pragma unroll
+      for (int64_t k = 0; k < nrows; ++k) {
+        acc[0][k] += load<scalar_t>(sum_base, col_stride, k);
+      }
+    }
+
+    for (int64_t j = 1; j < num_levels; ++j) {
+      #pragma unroll
+      for (int64_t k = 0; k < nrows; ++k) {
+        acc[j][k] += acc[j-1][k];
+        acc[j-1][k] = scalar_t(0);
+      }
+
+      const auto mask = (level_mask << (j * level_power));
+      if ((i & mask) != mask) {
+        break;
+      }
+    }
+  }
+
+  for (; i < size; ++i) {
+    const char * sum_base = in_data + i * row_stride;
+    #pragma unroll
+    for (int64_t k = 0; k < nrows; ++k) {
+      acc[0][k] += load<scalar_t>(sum_base, col_stride, k);
+    }
+  }
+
+  for (int64_t j = 1; j < num_levels; ++j) {
+    #pragma unroll
+    for (int64_t k = 0; k < nrows; ++k) {
+      acc[0][k] += acc[j][k];
+    }
+  }
+
+  std::array<scalar_t, nrows> ret;
+  for (int64_t k = 0; k < nrows; ++k) {
+    ret[k] = acc[0][k];
+  }
+  return ret;
+}
+
+template <typename scalar_t>
+scalar_t row_sum(const char * C10_RESTRICT in_data,
+                 const int64_t in_stride, const int64_t size) {
+  constexpr int64_t ilp_factor = 4;
```

Comment:
I notice `row_sum` is called with both a non-vectorized scalar type, and a vectorized scalar type. Does the ILP factor matter when you're not vectorizing? Why or why not?

Reply:
To the best of my knowledge, vectorised and scalar additions have roughly the same characteristics in terms of register pressure and instruction latency.  Both can be pipelined in the CPU hardware to achieve greater parallelism. I expect this is the main reason why my implementation is so much faster in non-vectorised cases, since there was no ILP before.